File: backend/pkg/handlers/web/memoryIssue.go
package web

import (
	"encoding/json"
	"log"
	"math"

	. "openreplay/backend/pkg/messages"
)

const MIN_COUNT = 3
const MEM_RATE_THRESHOLD = 300 // % to average

type MemoryIssueDetector struct {
	startMessageID uint64
	startTimestamp uint64
	rate           int
	count          float64
	sum            float64
	contextString  string
}

func (f *MemoryIssueDetector) reset() {
	f.startTimestamp = 0
	f.startMessageID = 0
	f.rate = 0
}

func (f *MemoryIssueDetector) Build() Message {
	if f.startTimestamp == 0 {
		return nil
	}
	payload, err := json.Marshal(struct{ Rate int }{f.rate - 100})
	if err != nil {
		log.Printf("can't marshal MemoryIssue payload to json: %s", err)
	}
	event := &IssueEvent{
		Type:          "memory",
		Timestamp:     f.startTimestamp,
		MessageID:     f.startMessageID,
		ContextString: f.contextString,
		Payload:       string(payload),
	}
	f.reset()
	return event
}

func (f *MemoryIssueDetector) Handle(message Message, timestamp uint64) Message {
	switch msg := message.(type) {
	case *PerformanceTrack:
		if f.count < MIN_COUNT {
			f.sum += float64(msg.UsedJSHeapSize)
			f.count++
			return nil
		}

		average := f.sum / f.count
		rate := int(math.Round(float64(msg.UsedJSHeapSize) / average * 100))

		f.sum += float64(msg.UsedJSHeapSize)
		f.count++

		if rate >= MEM_RATE_THRESHOLD {
			if f.startTimestamp == 0 {
				f.startTimestamp = timestamp
				f.startMessageID = message.MsgID()
			}
			if f.rate < rate {
				f.rate = rate
			}
		} else {
			return f.Build()
		}
	case *SetPageLocation:
		f.contextString = msg.URL
	}
	return nil
}

File: tracker/tracker/src/webworker/detectors/memoryIssue.ts
import Message, {
  Type,
  PerformanceTrack,
} from '../../common/messages.gen.js'
import type { Detector, ReportIssue } from './types.js'

/**
 * Port of backend/pkg/handlers/web/memoryIssue.go.
 *
 * Keeps a running average of used JS heap size and fires when the current
 * usage climbs to >= MEM_RATE_THRESHOLD % of that average. The running average
 * (sum/count) is intentionally NOT reset between issues.
 */

const MIN_COUNT = 3
const MEM_RATE_THRESHOLD = 300 // % of the running average

export default class MemoryIssueDetector implements Detector {
  readonly types: readonly number[] = [Type.PerformanceTrack]

  private startTimestamp = 0
  private startMessageId = 0
  private rate = 0
  private count = 0
  private sum = 0

  constructor(
    private readonly report: ReportIssue,
    private readonly url: () => string,
  ) {}

  private reset(): void {
    this.startTimestamp = 0
    this.startMessageId = 0
    this.rate = 0
  }

  private build(): void {
    if (this.startTimestamp === 0) {
      return
    }
    this.report({
      type: 'memory',
      contextString: this.url(),
      payload: JSON.stringify({ Rate: this.rate - 100 }),
      timestamp: this.startTimestamp,
      messageId: this.startMessageId,
    })
    this.reset()
  }

  flush(): void {
    this.build()
  }

  handle(message: Message, index: number, timestamp: number): void {
    const usedJSHeapSize = (message as PerformanceTrack)[4]

    if (this.count < MIN_COUNT) {
      this.sum += usedJSHeapSize
      this.count++
      return
    }

    const average = this.sum / this.count
    const rate = Math.round((usedJSHeapSize / average) * 100)

    this.sum += usedJSHeapSize
    this.count++

    if (rate >= MEM_RATE_THRESHOLD) {
      if (this.startTimestamp === 0) {
        this.startTimestamp = timestamp
        this.startMessageId = index
      }
      if (this.rate < rate) {
        this.rate = rate
      }
    } else {
      this.build()
    }
  }
}
